src/cache/cache_set.py

```diff
@@ -0,0 +1,275 @@
+            else:  # Current node is right child
+                self.state |= 1 << parent  # Set parent's bit
+            node = parent
+
+    def __get_plru_victim(self) -> int:
+        """
+        Find the way to replace according to PLRU policy.
+        Returns:
+            - Way index of victim line
+        """
+        node = 0
+        for _ in range(self.tree_levels):
+            if self.state & (1 << node):
+                node = 2 * node + 1  # Go left
+            else:
+                node = 2 * node + 2  # Go right
+        # Calculate leaf node index based on number of ways
+        return node - (self.num_ways - 1)
+
+    def __find_line(self, tag: int) -> Tuple[Optional[CacheLine], Optional[int]]:
+        """
+        Private Search Method: Find a cache line by tag.
+            Returns (line, way_index) or (None, None)
+        """
+        for i, line in enumerate(self.ways):
+            if not line.is_invalid() and line.tag == tag:
+                return line, i
+        return None, None
+
+    def read(self, tag: int) -> bool:
+        """
+        Simulates a read request to a cache set. If tag is found returns HIT, else MISS.
+        Upon a valid access plru bits are updated.
+        """
+        line, way_index = self.__find_line(tag)
+        if line is not None:  # Hit
+            # Update PLRU state bits on hit
+            self.__update_plru(way_index)
+            return True
+        return False  # Miss
+
+    def write(self, tag: int) -> bool:
+        """
+        Simulates a cache write request to a set. Write allocate policy is employed so any
+        write misses require cache to allocate line.
+        If tag is found returns HIT, else MISS.
+        """
+        line, way_index = self.__find_line(tag)
+        if line is not None:  # Hit
+            self.__update_plru(way_index)
+
+            return True
+        return False  # Miss
+
+    def allocate(
+        self, tag: int, state: MESIState = MESIState.EXCLUSIVE
+    ) -> Tuple[Optional[CacheLine], int]:
+        """
+        Allocate a new cache line. First searches for invalid lines
+        if none found uses PLRU to select victim.
+        Args:
+            tag: Tag to be stored in the cache line
+            state: Initial MESI state (default Exclusive for new allocations)
+
+        Returns:
+            (victim_line , allocated_way_index)
+            - victim_line: CacheLine object of the victim line
+            - If victim_line is None, no writeback needed (invalid line was used)
+            - If victim_line is Modified, writeback needed
+            - victim_line contains tag and L1 status for inclusion handling
+        """
+        # First check for invalid lines
+        for way_index, line in enumerate(self.ways):
+            if line.is_invalid():
+                # Found an invalid line, use it
+                line.tag = tag
+                line.mesi_state = state
+                line.in_l1 = False
```

Comment:
When / where do we update `in_L1` to `True`? Like the above, I understand this might not make sense for this PR, but I'm just trying to understand all the logic we have here so far.

Reply:
This is for inclusivity but I haven't thought of the exact logic for this yet. I'll add a todo comment here.